### ISSUE TITLE
XLA SPMD-partitioned FFT emits complex64 twiddle factors for a complex128 transform (HLO verifier rejects the mixed multiply)

### ISSUE THREAD

**Issue by Gordi42** [jax-ml/jax#39291]
### Description

A jitted complex128 FFT whose transform axis is sharded fails to
compile when the FFT result feeds a consumer that keeps that axis
sharded. XLA then lowers the FFT through its distributed Cooley-Tukey
decomposition, whose twiddle-factor constants come out as `complex64`
against the complex128 data, and the HLO verifier aborts compilation.

```
INVALID_ARGUMENT: during context [hlo verifier]: Binary op multiply with
different element types: c64[] and c128[].
, for instruction %multiply = c64[] multiply(%constant.1, %get-tuple-element.4),
metadata={op_name="jit(fft)/fft"}
```

The reproducer fails on jax 0.10.2 and 0.11.0, on CUDA and on the CPU
backend with forced host devices.

```python
import jax
jax.config.update("jax_enable_x64", True)
import jax.numpy as jnp
import numpy as np
from jax.sharding import Mesh, NamedSharding, PartitionSpec as P

n = 16
mesh = Mesh(np.array(jax.devices()), ("d",))
sh = lambda x, s: jax.device_put(x, NamedSharding(mesh, P(*s)))

x = sh(np.random.default_rng(0).standard_normal((n, n, n)), ("d", None, None))
q = jnp.asarray(np.random.default_rng(1).standard_normal((n, n, n, n)))

def f(a, q):
    c = jnp.fft.fft(a.astype(jnp.complex128), axis=0)   # norm=None; axis 0 sharded
    return jnp.einsum("...dj,...d->...j", q, jnp.moveaxis(c, 1, -1))

jax.jit(f)(x, q).block_until_ready()   # <-- c64/c128 HLO verifier error
```

run as

```
JAX_PLATFORMS=cuda python repro.py
# or without any GPU
XLA_FLAGS=--xla_force_host_platform_device_count=4 \
JAX_PLATFORMS=cpu python repro.py
```

A non-transform axis sharded (`P(None, "d", None)`) is fine.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.10.2
jaxlib: 0.10.2
numpy:  2.5.0
python: 3.12.11 (main, Aug  8 2025, 17:06:48) [Clang 20.1.4 ]
device info: NVIDIA A100-SXM4-80GB-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='l50003.lvt.dkrz.de', release='4.18.0-553.141.1.el8.x86_64', version='#1 SMP Thu Jul 9 20:00:37 CEST 2026', machine='x86_64')

$ nvidia-smi
Sat Jul 18 16:22:44 2026
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.159.04             Driver Version: 580.159.04     CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A100-SXM4-80GB          On  |   00000000:44:00.0 Off |                    0 |
| N/A   44C    P0             84W /  500W |    9637MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A         4024881      C   ...m/fridom-dev/.venv/bin/python       9196MiB |
|    0   N/A  N/A         4025449      C   .../fridom-dev/.venv/bin/python3        424MiB |
+-----------------------------------------------------------------------------------------+
```

**Comment by Saad-Mallebhari** [comment]
Hi @Gordi42

Thanks for the detailed issue report and the exact reproduction snippet.

Root Cause

In `jax/_src/fft.py` (and SPMD partitioner lowering for distributed FFTs in `jax/_src/interpreters/pxla.py`), the Cooley-Tukey decomposition pass generates twiddle factor constant arrays using a default `complex64` (`c64[]`) dtype instead of inheriting the input operand's precision (`complex128` / `c128[]`). When the SPMD partitioner emits the elementwise binary multiply between the `c64[]` twiddle constants and `c128[]` FFT data, the XLA HLO verifier rejects the mismatched instruction with an `INVALID_ARGUMENT` type mismatch error.

Proposed Solution

1. Refactor twiddle factor constant generation in `jax/_src/fft.py` (and SPMD FFT lowering passes) to dynamically match the input tensor's element type (`dtype = jnp.result_type(x)` or `complex128` when `jax_enable_x64` is active).
2. Ensure the Cooley-Tukey distributed lowering pass emits type-consistent binary multiplies (`c128[] multiply(c128[], c128[])`) when partitioning `complex128` FFT transforms.
3. Add a regression unit test in `tests/fft_test.py` asserting that sharded `complex128` 1D and 2D FFTs compile and execute without HLO verifier errors under CPU host device meshes (`xla_force_host_platform_device_count=4`).

Would you like me to open a Pull Request with this FFT lowering fix and accompanying regression tests?

**Comment by Gordi42** [comment]
This does not match the JAX source. There is no jax/_src/fft.py, the word twiddle appears nowhere in JAX's Python code, and pxla.py contains no FFT logic. The bad constant is emitted by XLA's C++ SPMD partitioner, so no dtype change on the JAX side would fix it.

The actual fix already landed in openxla/xla#46450 and closes openxla/xla#44541, which is the same bug filed a month before this one. Closing as fixed upstream.
